### PR TITLE
GA-55 process 2

### PR DESCRIPTION
**Pull request by abul-syed**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/GA-55

### Change description ###

PRM : Process 2 - Find Organisations with Stale Profiles

### Updates

Update to entities (add `retry` & `retry-after` columns)

- #1772
- #1773 

Add `organisation_last_updated` - this is last_updated from PRD
Add `last_updated` - will be used for queue clean up process to delete record last_updated > (now - 90 days)

- #1791

### CODE REVIEW

File: src/contractTest/java/uk/gov/hmcts/reform/refdata/RefDataProfessionalOrganisationalInternalConsumerTest.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                                .build()
+                        ))
+                .method("POST")
+                .willRespondWith()
+                .status(200)
+                .body(buildOrganisationResponsePactDsl())
+                .toPact();
+    }
+
+    @Test
```
Comment (sanjayparekh-hmcts) [inline src/contractTest/java/uk/gov/hmcts/reform/refdata/RefDataProfessionalOrganisationalInternalConsumerTest.java:83]: I think we've decided to disable these pats for PR runs for now

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        organisationService.findAndInsertStaleOrganisationsIntoRefreshQueue();
+
+        assertEquals(organisationRefreshQueueRepository.findAll().size(), 1);
+        assertEquals(profileRefreshQueueRepository.getActiveProfileEntities().size(), 0);
+    }
+
+    @Test
+    @Sql(executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD,
+            scripts = {"classpath:sql/insert_profile_refresh_queue.sql"})
+    void shouldInsertTwoOrganisationIntoOrganisationRefreshQueue_AndClearProfileRefreshQueue() {
```
Comment (sanjayparekh-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceIntegrationTest.java:65]: I wonder if this test is overkill? Is there any DB integrity you are checking by having 2 orgs as opposed to 1 org?

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -186,3 +188,12 @@ springdoc:
...
   writer-with-order-by-keys: true
+
+professional:
+  refdata:
+      pageSize: ${PROFESSIONAL_REFDATA_PAGE_SIZE:100}
+  role:
+    mapping:
+      scheduling:
+        enabled: ${PROFESSIONAL_SCHEDULING_ENABLED:false}
+        cron: ${PROFESSIONAL_ROLE_MAPPING_CRON:"*/30 * * * * *"}
```
Comment (sanjayparekh-hmcts) [inline src/main/resources/application.yaml:199]: Dan and Suman have just determined that we will have multiple schedules in crons, so we probably want this value to be specific per Process

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/scheduler/Scheduler.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+@Service
+public class Scheduler {
+
+    private final OrganisationService organisationService;
+
+    public Scheduler(OrganisationService organisationService) {
+        this.organisationService = organisationService;
+    }
+
+    @Scheduled(cron = "${professional.role.mapping.scheduling.cron}")
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/scheduler/Scheduler.java:16]: see below

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/PrdService.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import uk.gov.hmcts.reform.orgrolemapping.domain.model.OrganisationByProfileIdsResponse;
+import uk.gov.hmcts.reform.orgrolemapping.feignclients.PRDFeignClient;
+
+@Service
+@AllArgsConstructor
+public class PrdService {
+
+    private final PRDFeignClient prdFeignClient;
+
+    @Retryable(maxAttempts = 3, backoff = @Backoff(delay = 500, multiplier = 3))
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/PrdService.java:18]: Dan L has OKd this, but suggested we keep an eye on it esp during PerfTest

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationService.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    private void updateOrganisationRefreshQueueActiveStatus(List<String> organisationProfileIds,
+                                                            Integer accessTypeMaxVersion) {
+        organisationProfileIds.forEach(organisationProfileId -> profileRefreshQueueRepository.setActiveFalse(
+                organisationProfileId,
+                accessTypeMaxVersion
+        ));
+    }
+
+    private boolean responseNotNull(OrganisationByProfileIdsResponse response) {
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationService.java:102]: Last I looked there were some missing code coverage on the usage of this method

File: src/contractTest/java/uk/gov/hmcts/reform/refdata/RefDataProfessionalOrganisationalInternalConsumerTest.java (hunk 1 above)
Comment (abul-syed) [reply src/contractTest/java/uk/gov/hmcts/reform/refdata/RefDataProfessionalOrganisationalInternalConsumerTest.java:86]: disabled in commit [9375d38](https://github.com/hmcts/am-org-role-mapping-service/pull/1743/commits/9375d389ba6ce2749085574411c4ea75aa7fd2f6)

File: src/main/resources/application.yaml (hunk 3 above)
Comment (abul-syed) [reply src/main/resources/application.yaml]: updated in [9375d38](https://github.com/hmcts/am-org-role-mapping-service/pull/1743/commits/9375d389ba6ce2749085574411c4ea75aa7fd2f6)

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/scheduler/Scheduler.java (hunk 4 above)
Comment (abul-syed) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/scheduler/Scheduler.java]: updated in [9375d38](https://github.com/hmcts/am-org-role-mapping-service/pull/1743/commits/9375d389ba6ce2749085574411c4ea75aa7fd2f6)

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationService.java (hunk 6 above)
Comment (abul-syed) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationService.java:102]: Screenshot 2024-02-09 at 14 55 22

code code now at 100% for service.

i have removed `&& maxVersion.isPresent()` - as the check is redundant, as it's always `true` by the time the code hits the `if` statement. 

also modified `responseNotNull` method to just be `return response != null && !response.getOrganisationInfo().isEmpty();` these checks are sufficient so no need for additional `!response.getLastRecordInPage().isEmpty()
                && response.getMoreAvailable() != null` check

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceIntegrationTest.java (hunk 2 above)
Comment (abul-syed) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceIntegrationTest.java]: have removed IT in commit [d877fc1](https://github.com/hmcts/am-org-role-mapping-service/pull/1743/commits/d877fc137bdf20ec1516907aacb376c4e7142fca)

**Review by sanjayparekh-hmcts — APPROVED** [review]

**Review by sanjayparekh-hmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/model/OrganisationInfo.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.NoArgsConstructor;
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class OrganisationInfo {
```
Comment (shaed-parkar) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/domain/model/OrganisationInfo.java:15]: Is there any means of packaging the contract models from Ref Data?

**Review by shaed-parkar — APPROVED** [review]
Looks good. You might want to hold off on merging until we confirm the mandatory query params

**Review by prashanthkumarnb — APPROVED** [review]

**Review by mattnayler — APPROVED** [review]
Approved with a couple of comments that would be nice to haves in a future PR.

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepositoryIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        assertNotNull(organisationEntity.getOrganisationLastUpdated());
+        assertNotNull(organisationEntity.getLastUpdated());
+        assertNotNull(organisationEntity.getRetryAfter());
+    }
+
+    @Test
+    @Sql(executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD, scripts = {
+        "classpath:sql/insert_organisation_profiles.sql"
+    })
+    public void shouldHandleInsertConflictIntoOrganisationRefreshQueue() {
```
[the remark this answers is not in the document]
Comment (shaed-parkar) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepositoryIntegrationTest.java:47]: added it :) good shout

File: src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                .thenReturn(ResponseEntity.ok(response));
+
+        organisationService.findAndInsertStaleOrganisationsIntoRefreshQueue();
+
+        verify(profileRefreshQueueRepository, times(1))
+                .getActiveProfileEntities();
+        verify(organisationRefreshQueueRepository, times(1))
+                .upsertToOrganisationRefreshQueue(any(), any(), any());
+        verify(profileRefreshQueueRepository, times(1))
+                .setActiveFalse(any(), any());
```
[the remark this answers is not in the document]
Comment (shaed-parkar) [reply src/test/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/OrganisationServiceTest.java:68]: you should find the tests in the `OrganisationRefreshQueueRepositoryIntegrationTest` should satisfy the the assertions you would like to see

**Pull request merged**